lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        key = Rails.application.key_generator.generate_key("discourse-auth-cookie", 32)
+        ActiveSupport::MessageEncryptor.new(
+          key,
+          key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
+        )
+      end
+    end
+  end
+
+  TOKEN_SIZE ||= 32
+
+  TOKEN_KEY ||= "token"
+  ID_KEY ||= "id"
+  TL_KEY ||= "tl"
+  TIME_KEY ||= "time"
+  VALID_KEY ||= "valid"
```

Comment:
```suggestion
  TOKEN_SIZE = 32

  TOKEN_KEY = "token"
  ID_KEY = "id"
  TL_KEY = "tl"
  TIME_KEY = "time"
  VALID_KEY = "valid"
```

We don't need `||=` when assigning constants. We used to do it historically but that is because we had a bug in the application that would try to load a file twice.

Reply:
I think I saw constant redefined warnings printed the console while I was working on this which led me to adding them, but will confirm.

Reply:
Those redefined warnings indicate a bug which we should fix :grin: